🧪 Moonwall - DevTest Migration

timbrinded opened the pull request:
### What does it do?
First attempt to migration our extensive dev tests (aka Typescript Tests) over to moonwall.
This fleshes out the tests in the `/test` subdirectory and includes:
- :heavy_plus_sign: Adds ~~foundry/forge~~ new script to the toolchain for contract compilation
- :heavy_plus_sign: Adds about 60% of typescript dev tests to moonwall
- :heavy_plus_sign: Adds a new CI automation `build/dev-test`
- :heavy_division_sign: Small fixes to moonwall smoke-tests
- :heavy_division_sign: Updates to moonwall deps
- :heavy_minus_sign: Removed existing compiled contracts directory from `/test` (the compiled .jsons are unaffected for existing `/tests` directory)

Currently this runs multi-core and takes a few minutes to run:
![image](https://github.com/PureStake/moonbeam/assets/79199034/d6b08ac3-4631-449b-9244-ea26e2f31c90)

### What important points reviewers should know?
Apologies for the 274 files changed, many of that will be adding solidity contracts to our new foundry subdir, and removing the compiled jsons from the existing helpers subdir. I want to get away from a large helpers directory being a permanent store of logic, and mature code being genericized and moved into the moonwall util repo.

### Is there something left for follow-up PRs?
- Rest of Dev tests packs to be migrated over
- Brand new PoV tests

### What value does it bring to the blockchain users?
Better monitoring of the live blockchains, and a more efficient and future-proof platform to write new tests that removes existing tech debt.

crystalin commented on test/contracts/src/CallPermitDemo.sol:
```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity >=0.8.3;
+
+import "precompiles/call-permit/CallPermit.sol";
```
Have you tried loading it in Remix ? We used to have relative path to allow tools like VScode and Remix to access the file.
(same comment for all other solidity files using precompiles)

crystalin commented on test/helpers/assets.ts:
```diff
@@ -0,0 +1,172 @@
+  expect(assets.unwrap().asXcm.parents.toNumber()).to.equal(1);
+
+  // Get keys to modify balance
+  let module = xxhashAsU8a(new TextEncoder().encode("Assets"), 128);
```
I'd suggest to use "const" keyword for immutable variables. This makes it easier to read

crystalin commented on test/helpers/block.ts:
```diff
@@ -0,0 +1,378 @@
+  txWithEvents: TxWithEventAndFee[];
+}
+
+const getBlockDetails = async (
```
I think the functions in this file should definitely go to moonwall somehow and be more generic.

crystalin commented on test/helpers/block.ts:
```diff
@@ -0,0 +1,378 @@
+  }
+};
+
+export const verifyBlockFees = async (
```
This could get moved to moonbeam-tools or something specific, as it has been copied to few places and is not maintained in all of them

crystalin commented on test/helpers/block.ts:
```diff
@@ -0,0 +1,378 @@
+  }
+}
+
+export async function jumpRounds(context: DevModeContext, count: Number): Promise<string | null> {
```
In moonwall, with maybe support for different round meaning based on which pallet is available. Just an idea

timbrinded replied:
Good idea, raised it at https://github.com/Moonsong-Labs/moonwall/issues/140 which will be done for future PR when i cover parachain staking tests

crystalin commented on test/helpers/common.ts:
```diff
@@ -18,9 +30,96 @@ export async function checkTimeSliceForUpgrades(
+ * @param {Record<string, any>} unsortedObject - The object with unsorted keys.
+ * @returns {Record<string, any>} - The new object with keys sorted.
+ */
+export function sortObjectByKeys(unsortedObject: Record<string, any>): Record<string, any> {
```
Maybe we should use a standard library for small useful functions instead of writing our own.

crystalin commented on test/package.json:
```diff
@@ -14,38 +14,47 @@
+    "tsconfig-paths": "^4.2.0",
+    "viem": "^0.3.50",
+    "vitest": "^0.31.4",
+    "web": "^0.0.2"
```
what is this ?

timbrinded replied:
done

crystalin commented on test/suites/rt-upgrade/test-upgrade-chain.ts:
```diff
@@ -45,7 +45,7 @@ describeSuite({
         const balanceBefore = (await api.query.system.account(DUMMY_ACCOUNT)).data.free.toBigInt();
         await api.tx.balances.transfer(DUMMY_ACCOUNT, parseEther("1")).signAndSend(alith);
-        await context.createBlock();
+        await context.createBlock({ count: 2 });
```
Why 2 blocks ?

timbrinded replied:
Because chopsticks has been randomly failing in other [CI runs](https://github.com/PureStake/moonbeam/actions/runs/5210242684/jobs/9401534547?pr=2332), which i suspect is a racecondition. I figured generating an extra block should ensure the state is properly updated.

crystalin commented on tests/smoke-tests/test-relay-indices.ts:
```diff
@@ -107,6 +107,7 @@ describeSmokeSuite(
 
     testIt("C400", "should have matching indices for Staking.Bond", async function () {
       const callHex = context.relayApi.tx.staking
+        // @ts-ignore
```
Why are those needed ?

timbrinded replied:
Get ts errors due to ApiAugment having mismatching function signatures. E.g.:
```
 staking: {
            /**
             * See [`Pallet::bond`].
             **/
            bond: AugmentedSubmittable<(value: Compact<u128> | AnyNumber | Uint8Array, payee: PalletStakingRewardDestination | {
                Staked: any;
            } | {
                Stash: any;
            } | {
                Controller: any;
            } | {
                Account: any;
            } | {
                None: any;
            } | string | Uint8Array) => SubmittableExtrinsic<ApiType>, [Compact<u128>, PalletStakingRewardDestination]>;
```
2 args, not 3. 
Will investigate further given that polkadot.js shows the fun as expected (maybe its an api-augment issue only).

timbrinded replied:
This is a result of function signatures changing in Polkadot RT9430, which won't be for a while.

Have changed this to @ts-expect-errors for now, and will remove when we need to update our fn signatures

timbrinded commented on .github/workflows/build.yml:
```diff
@@ -588,3 +589,50 @@ jobs:
+    needs: ["set-tags", "build"]
+    strategy:
+      matrix:
+        chain: ["moonbase"]
```
In the future, yes! But saving for future PR.

Given our three chains can have three different feature-sets at a given time, we need to design a three tiered structure that's maintainable.

timbrinded commented on .github/workflows/build.yml:
```diff
@@ -588,3 +589,50 @@ jobs:
+          cd test
+          pnpm i
+          pnpm moonwall test dev_${{ matrix.chain }}
+      - name: Upload HTML report
```
Yes, that would be great actually. 
Maybe only needed for failures though (to save on space and time)

timbrinded commented on test/.gitmodules:
I've removed it, but in future will be slightly more annoying to update our foundry and openzepplin versions if we wanted to. (solc ver management however is unaffected and still easy)

timbrinded commented on .gitmodules:
```diff
@@ -0,0 +1,6 @@
+[submodule "test/contracts/lib/forge-std"]
```
Killed it 🪓

timbrinded commented on .github/workflows/build.yml:
```diff
@@ -588,3 +589,50 @@ jobs:
           chmod uog+x tmp/polkadot
           pnpm moonwall test zombie_${{ matrix.chain }}_ci
+
+  dev-test:
```
done

timbrinded commented:
@crystalin PR comments addressed, namely we are now back down to *only* 277 files changed.

- Added a compile script for smart contracts
- Added moonwall config to run this script on every dev test run
- Removed foundry files and libs
- Updated package dependency vers
